src/utils/TodoContext.tsx

```diff
@@ -0,0 +1,205 @@
+
+    sendTodoToServer(newTodo)
+      .then(response => {
+        setTodos(prevTodos => [...prevTodos, response]);
+        setQuery('');
+      })
+      .catch(() => {
+        setError(ErrorMessage.ADD_TODO_ERROR);
+      })
+      .finally(() => {
+        setTempTodo(null);
+        setIsLoading(false);
+        setLoadingTodosIDs(prev => prev.filter(todoId => todoId !== 0));
+      });
+  };
+
+  const removeTodo = (id: number) => {
+    setIsLoading(true);
+    setLoadingTodosIDs(prev => [...prev, id]);
+    deleteTodoFromServer(id)
+      .then(() => {
+        setTodos(prevTodos => prevTodos.filter(todo => todo.id !== id));
+      })
+      .catch(() => {
+        setError(ErrorMessage.DELETE_TODO_ERROR);
+      })
+      .finally(() => {
+        setIsLoading(false);
+        setLoadingTodosIDs(prev => prev.filter(todoId => todoId !== id));
+      });
+  };
+
+  const toggleTodoStatus = (id: number, status: boolean) => {
+    setIsLoading(true);
+    setLoadingTodosIDs(prev => [...prev, id]);
+
+    const todoToUpdate = todos.find(todo => todo.id === id);
+
+    if (todoToUpdate) {
+      toggleStatusTodo(id, status)
+        .then(() => {
+          setTodos(prevTodos =>
+            prevTodos.map(prev =>
+              prev.id === id ? { ...prev, completed: status } : prev,
+            ),
+          );
+        })
+        .catch(() => {
+          setError(ErrorMessage.UPDATE_TODO_ERROR);
+        })
+        .finally(() => {
+          setIsLoading(false);
+          setLoadingTodosIDs(prev => prev.filter(todoId => todoId !== id));
+        });
+    }
+  };
+
+  const toggleTodo = (id: number) => {
+    const todoToUpdate = todos.find(todo => todo.id === id);
+
+    if (todoToUpdate) {
+      toggleTodoStatus(id, !todoToUpdate.completed);
+    }
+  };
+
+  const toggleAllTodos = (status: boolean) => {
+    todos.forEach(todo => {
+      if (todo.completed !== status) {
+        toggleTodoStatus(todo.id, status);
+      }
+    });
+  };
+
+  const renameTodo = (
+    id: number,
+    newTitle: string,
+    setEditingTodoId: (id: number | null) => {},
+  ) => {
+    setIsLoading(true);
+    setLoadingTodosIDs(prev => [...prev, id]);
+
+    renameTodoOnServer(id, newTitle)
+      .then(() => {
+        setTodos(prevTodos =>
+          prevTodos.map(prev =>
+            prev.id === id ? { ...prev, title: newTitle } : prev,
+          ),
+        );
+        setEditingTodoId(null);
+      })
+      .catch(() => {
+        setError(ErrorMessage.UPDATE_TODO_ERROR);
+      })
+      .finally(() => {
+        setIsLoading(false);
+        setLoadingTodosIDs(prev => prev.filter(todoId => todoId !== id));
+      });
+  };
+
+  useEffect(() => {
+    const fetchTodos = async () => {
+      setIsLoading(true);
+      try {
+        const fetchedTodos = await getTodos();
+
+        setTodos(fetchedTodos);
+      } catch {
+        setError(ErrorMessage.UNABLE_LOAD_TODOS);
+      } finally {
+        setIsLoading(false);
+      }
+    };
```

Comment:
Let's have some consistency, if you decided to use `async/await` use it everywhere